### ISSUE TITLE
bug: `appium driver list --installed` is unable to find drivers in monorepo

### ISSUE THREAD

**Issue by tido64** [appium/appium#19819]
### Do I have the most recent component updates?

- [X] I use the most recent available driver/plugin and server versions

### Is the component officially supported by the Appium team?

- [X] I have verified the component repository is present under the Appium organization in GitHub

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

`appium driver list --installed` lists no drivers in a monorepo when drivers are hoisted to the root workspace. Example:

```
monorepo
├── node_modules
│   ├── appium
│   ├── appium-uiautomator2-driver  <-- drivers were installed here
│   └── appium-xcuitest-driver      <-- drivers were installed here
├── package.json
└── packages
    └── my-app
        ├── node_modules  <-- no drivers here; they were hoisted to root
        ├── package.json  <-- drivers were requested here
        └── src
```

### Expected Behavior

`appium driver list --installed` should list installed drivers

### Minimal Reproducible Example

If you have drivers installed globally, move them elsewhere before running the following commands:

```
git clone https://github.com/microsoft/react-native-test-app.git
cd react-native-test-app
git reset --hard 329f07c5e580c792b549b9a19c90f7e79f4f062c
yarn
cd example
appium driver list --installed
```

### Environment

- Operating system: macOS 14.3.1
- Appium server version (output of `appium --version`): 2.5.1
- Appium driver(s) and their version(s):
  - appium-uiautomator2-driver 3.0.0
  - appium-xcuitest-driver 7.1.0
- Appium plugin(s) and their version(s): n/a
- Node.js version (output of `node --version`): v18.19.1
- `npm` version (output of `npm --version`): 10.2.4
- Last component(s) version which did _not_ exhibit the problem: n/a
- Platform and version under test: ?
- Real device or emulator/simulator: emulator/simulator

### Link to Appium Logs

n/a

### Further Information

We've worked around this issue with the following patch:

```diff
diff --git a/build/lib/extension/manifest.js b/build/lib/extension/manifest.js
index 5b650a6ff4dbe737ffdf8821e255e9e4e3d89da3..6179965231285c3ae1d3cddee0c45c57ab3fba03 100644
--- a/build/lib/extension/manifest.js
+++ b/build/lib/extension/manifest.js
@@ -195,12 +195,14 @@ class Manifest {
             onMatch(path_1.default.join(__classPrivateFieldGet(this, _Manifest_appiumHome, "f"), 'package.json'), true),
         ];
         // add dependencies to the queue
-        const filepaths = await support_1.fs.glob('node_modules/{*,@*/*}/package.json', {
-            cwd: __classPrivateFieldGet(this, _Manifest_appiumHome, "f"),
-            absolute: true,
-        });
-        for (const filepath of filepaths) {
-            queue.push(onMatch(filepath));
+        const appPkgPath = path_1.default.join(process.cwd(), 'package.json');
+        const appPkg = JSON.parse(await support_1.fs.readFile(appPkgPath, 'utf8'));
+        for (const name of Object.keys(appPkg.devDependencies)) {
+          try {
+            queue.push(onMatch(require.resolve(name + "/package.json")));
+          } catch (_) {
+            // Ignore packages that don't export `./package.json`
+          }
         }
         // wait for everything to finish
         await bluebird_1.default.all(queue);
```

**Comment by jlipps** [comment]
I wonder if we should add something like a `--ext-search-root` flag where you could simply tell Appium the directory to start its hunt for extension packages. That way you could simply run `appium --ext-search-root=$MONOREPO_ROOT`

**Comment by tido64** [comment]
With hoisting in general, I don't think you can reliably set a single path for globbing. Since it's based on popularity, you never really know when the next set of package updates might move things around. You will have to check both the current workspace as well as the root workspace. I'm also not sure globbing is the best solution here since you might find extensions that the current workspace don't have an explicit dependency on.

**Comment by jlipps** [comment]
Yes, fair point. I think what I meant was to use an approach like you do here, but instead of relying on `process.cwd()`, allowing the user to specify a path.

**Comment by szarlatan** [comment]
@tido64, I had pretty much the same problem and solution for that was quite simple with just a slight overhead.

Please try to add drivers' packages also as a `devDependencies` in your root monorepro `package.json` file. That solved the problem for me but each time I bump drivers versions I have to update 2 package.json files instead of one. Still that's a low cost IMO.

**Comment by trippingtarballs** [comment]
Beyond driver-package resolution, there appears to be other ways in which `appium` fails to function gracefully when included in a monorepo.

![image](https://github.com/user-attachments/assets/7828e3ac-a381-4ad9-aa0e-08e70fa39f26)

There is a lot of prior-art in this area. Please advise if support is needed to unpick these issues, I'd be happy to help.

**Comment by mykola-mokhnach** [comment]
@trippingtarballs  Of course we'd be happy to accept any help there. Just create a PR and we'll review it. You may also use Appium slack forum to ask any questions related to the development.

**Comment by MadeinFrance** [comment]
> Beyond driver-package resolution, there appears to be other ways in which `appium` fails to function gracefully when included in a monorepo.
> 
> ![image](https://private-user-images.githubusercontent.com/2758507/353505244-7828e3ac-a381-4ad9-aa0e-08e70fa39f26.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM2MTkyMzcsIm5iZiI6MTcyMzYxODkzNywicGF0aCI6Ii8yNzU4NTA3LzM1MzUwNTI0NC03ODI4ZTNhYy1hMzgxLTRhZDktYWEwZS0wOGU3MGZhMzlmMjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTRUMDcwMjE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmVjNzFiOWNhNGJiNzk1MGE4ZWVlMGUxZWY1YmM5ZTI1MmU3MmEwZGI4Yjc0YjYyNmMwNDUwYTI1Y2Q3OTU1ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.XDLvdBndz2vSmWLZFqVcwmW9s2ASjPAeEI9itZIgLXg)
> 
> There is a lot of prior-art in this area. Please advise if support is needed to unpick these issues, I'd be happy to help.

Adding Appium + drivers to the `devDepencies` didn't solve the issue. See the exact error:
`yarn appium driver install xcuitest@7.24.12;`
```
STDERR:
npm error code EUNSUPPORTEDPROTOCOL
npm error Unsupported URL Type "patch:": patch:my-rn-patch.patch"
```

@jlipps how to proceed? The proposed patch by @tido64 is fixing the issue.

**Comment by jlipps** [comment]
@trippingtarballs do you have any desire to propose a solution to this issue?
